cornell-dti/course-plan
# Two-graph requirement header progress colors and warnings

### Summary 

This pull request continues the two-graph front-end by updating the progress counter/bar/warning for college/major/minor headers. Adapted from @SamChou19815 's #599 and designers [here](https://www.figma.com/file/vz9T397rHdewbB2McmUDEr/Yuxuan-Working-File?node-id=2035%3A1560).

- [x] Set progress counter to represent all fulfilled requirements, even those that are dangerously fulfilled
- [x] Set progress bar color to be blue for safely fulfilled reqs, yellow for dangerously fulfilled
- [x] Add a `ProgressBarWarning` next to the bar to show number of conflicts when they exist. 
- [x] **NEW** Add GK so that yellow progress and new warnings only show when conflicts enabled 



Remaining TODOs:

- [x] Fix warning z-index issue, see Notes.
- [ ] Update CourseCaution warning icon color after #621 merged depending, see Notes.
- [ ] Add the functionality that actually creates conflicted requirements in a future PR.



Depends on #621 for full requirement bar two-graph implementation.

### Test Plan 

To test the below, I updated the function getAllCoursesThatCanPotentiallySatisfyRequirement in requirement-graph-builder-from-user-data.ts locally by adding an edge between a course and a new req in `dangerousRequirementFulfillmentGraph`, and removing the existing edge between that course and the edge it is assigned to in `safeRequirementFulfillmentGraph.` This simulates as if a course has two dangerous requirements it is being double counted for (and thus no safe requirement). To do so, the courseID and requirement name must be hardcoded in (accessed from Firestore and the Requirements debugger respectively).

1. Add a second fulfillment to the CS 4121 course in plan of a CS Major.
![image](https://user-images.githubusercontent.com/25535093/146882812-88ff78c6-53b4-4a9f-bd09-c7764b9a5745.png)

2. Confirm that then that the progress bar will be yellow, representing the practicum requirement has a conflict. Further confirm the fulfillment number / 8 still counts the practicum requirement being complete despite the conflict. Finally, confirm the warning icon with tooltip shows up next to the fulfillment count.
![image](https://user-images.githubusercontent.com/25535093/146882032-dc0e00de-5b64-4401-b6c7-635587b851b6.png)

3. Confirm the above also works for compounded requirements (i.e. liberal studies)
![image](https://user-images.githubusercontent.com/25535093/146882826-9094979f-8cfd-4bc2-8b8f-d136818b817f.png)
![image](https://user-images.githubusercontent.com/25535093/146882608-c02ecbce-4093-4ed7-916e-1184157766a8.png)

4. **NEW** Confirm that turning on/off GK turns off/on yellow progress bar + warning
![image](https://user-images.githubusercontent.com/25535093/160704555-35523f54-de9d-402a-bedb-d205f341768b.png)
![image](https://user-images.githubusercontent.com/25535093/160704704-0679f342-7bf5-4c0e-8d5e-67f66ad47371.png)

### Notes 

Two design/implementation questions
1. As seen in the screenshots above, the warning tooltip is partially covered by the `SemesterView`, especially on certain breakpoints. Changing the z-index does not seem to easily fix the issue. Any suggestions on how to fix this in the code, or if not, could the warning be placed to the left of `0/8 Total Requirements Inputted on Schedule` instead of to the right?
2. The warning icons for these new conflicts on Figma are a different color than our existing warnings (the new ones match the progress bar color & highlight color we use for conflicts). If this is intentional, should this color also be used on course cards that have conflicts? If so, what if a course has both a conflict and an existing type of warning (like duplicate)?

Once these issues are resolved, this PR can be safely merged without affecting users until conflicts can actually be created.
